packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/ScatterPlotViewer/components/ZoomingScatterPlot/ZoomingScatterPlot.js

```diff
@@ -0,0 +1,245 @@
+    const yScale = transformYScale(data, transformMatrix, this.rangeParameters)
+    const yScaleDomain = yScale.domain()
+    const outOfYAxisDataBounds = yScaleDomain[0] < (dataExtent.y[0] - PAN_DISTANCE) || yScaleDomain[1] > (dataExtent.y[1] + PAN_DISTANCE)
+
+    return outOfYAxisDataBounds
+  }
+
+  isYScaleMin(scaleY) {
+    const { minZoom } = this.props.zoomConfiguration
+
+    return scaleY < minZoom
+  }
+
+  isYScaleMax(scaleY) {
+    const { maxZoom } = this.props.zoomConfiguration
+
+    return scaleY > maxZoom
+  }
+
+  constrainXAxisZoom (transformMatrix, prevTransformMatrix) {
+    const { maxZoom } = this.props.zoomConfiguration
+    const { scaleX } = transformMatrix
+    const isXAxisOutOfBounds = this.isXAxisOutOfBounds(transformMatrix)
+    const isXScaleMin = this.isXScaleMin(scaleX)
+    const isXScaleMax = this.isXScaleMax(scaleX)
+
+    const newTransformMatrix = Object.assign({}, transformMatrix, { scaleY: 1, translateY: 0 })
+
+    if (isXScaleMin) {
+      return this.zoomReset()
+    }
+
+    if (isXScaleMax) {
+      newTransformMatrix.scaleX = maxZoom
+      newTransformMatrix.translateX = prevTransformMatrix.translateX
+    }
+
+    if (isXAxisOutOfBounds) {
+      return prevTransformMatrix
+    }
+
+    return newTransformMatrix
+  }
+
+  constrainYAxisZoom(transformMatrix, prevTransformMatrix) {
+    const { maxZoom } = this.props.zoomConfiguration
+    const { scaleY } = transformMatrix
+    const isYAxisOutOfBounds = this.isYAxisOutOfBounds(transformMatrix)
+    const isYScaleMin = this.isYScaleMin(scaleY)
+    const isYScaleMax = this.isYScaleMax(scaleY)
+
+    const newTransformMatrix = Object.assign({}, transformMatrix, { scaleX: 1, translateX: 0 })
+
+    if (isYScaleMin) {
+      return this.zoomReset()
+    }
+
+    if (isYScaleMax) {
+      newTransformMatrix.scaleY = maxZoom
+      newTransformMatrix.translateY = prevTransformMatrix.translateY
+    }
+
+    if (isYAxisOutOfBounds) {
+      return prevTransformMatrix
+    }
+
+    return newTransformMatrix
+  }
+
+  constrainBothAxisZoom(transformMatrix, prevTransformMatrix) {
+    const { maxZoom } = this.props.zoomConfiguration
+    const { scaleX, scaleY } = transformMatrix
+    const isXAxisOutOfBounds = this.isXAxisOutOfBounds(transformMatrix)
+    const isYAxisOutOfBounds = this.isYAxisOutOfBounds(transformMatrix)
+    const isXScaleMin = this.isXScaleMin(scaleX)
+    const isYScaleMin = this.isYScaleMin(scaleY)
+    const isXScaleMax = this.isXScaleMax(scaleX)
+    const isYScaleMax = this.isYScaleMax(scaleY)
+
+    if (isXScaleMax && isYScaleMax) {
+      return Object.assign({}, transformMatrix, {
+        scaleX: maxZoom,
+        scaleY: maxZoom,
+        translateX: prevTransformMatrix.translateX,
+        translateY: prevTransformMatrix.translateY
+      })
+    }
+
+    if (isXScaleMin || isYScaleMin) {
+      return this.zoomReset()
+    }
+
+    if (isXAxisOutOfBounds || isYAxisOutOfBounds) {
+      return prevTransformMatrix
```

Comment:
This implementation of the _"prevent user from panning beyond the bounds of the Subject's data"_ logic is legit, and is in fact too legit, too legit to quit.

Also, bonus 👍 for accounting for the wacky logic for x-axis-only constraints, y-axis-only constraints, and both-axis constraints.

One quirk to note is that using "revert to a previous transform" logic - as opposed to the more complex "calculate a new transform that is as far as the user wants to go but still within the bounds of the data" - will cause huge pan values to cancel out, leaving the user in the same place.

e.g. the valid data range is x=[0, 10], and currently x=2. If the user attempts to pan left by -5 the view will remain at x=2 instead of going "as far left as possible" to x=0.

That said, at the moment, this is only _a theoretical problem,_ as the only way I could "cancel my pan action", so to speak, was to purposely attempt to make a huge pan jump by quickly click-dragging a large distance - _and nobody sensible will do that._

This _may_ only matter if we introduce 'fixed' jumps later though, e.g keyboard left button jumps left by a fixed -5, but the view is at x=2, so pressing left key seems to "stop panning" well before the user sees the left-most edge of the Subject data.

(Clarification: this is not a blocker, but just a flag for possible future gotchas.)

Reply:
I'm not sure the end user will know that they aren't actually fully panned to the allowed range, so as long as the data is available in view since I'm using a constant to add a buffer similar to the LCV, I think it is ok. 

Of course, when we actually test this out with volunteers, my assumption may be proved wrong and I think you're correct that this might have to be adjusted when keyboard control is implemented too.